### PR TITLE
Output role cert key file

### PR DESCRIPTION
**Pull request by WindzCUHK**
# Description

issue: when external private key is rotated, the issued role certificate cannot be used

## Changes

- new config
  - ENV: `ROLE_CERT_KEY_FILE_OUTPUT`
  - args: `rolecert-key-file-output`
- output role cert key file alongside with the corresponding role cert
  - `rolecerts/domain_role.key.pem`
- update `go.mod`

## Type of change

- [x] Bug fix
- [x] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [x] Passed all pipeline checking

## Checklist for maintainer
- Use `Squash and merge`
- Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- Delete the branch after merge

### CODE REVIEW

**Comment by mlajkim** [comment]
Is v3.2.0 officially released after merging this PR?
@WindzCUHK

**Comment by WindzCUHK** [comment]
@mlajkim
> Is v3.2.0 officially released after merging this PR?
Yes, I think so.

File: pkg/identity/certificated.go [hunk 1]
```diff
@@ -102,17 +102,26 @@ func Certificated(idConfig *config.IdentityConfig, stopChan <-chan struct{}) (er
 					if err := w.AddBytes(outPath, 0644, roleCertPEM); err != nil {
 						return errors.Wrap(err, "unable to save x509 role cert")
 					}
+
+					// always output role cert key file to prevent unexpected key rotation when using external key
+					if id == nil || idConfig.RoleCertKeyFileOutput {
```
Comment (mlajkim) [inline pkg/identity/certificated.go:107]:
```suggestion
					if id != nil || idConfig.RoleCertKeyFileOutput {
```
Might be wrong but isn't it `id == nil`

Comment (mlajkim) [inline pkg/identity/certificated.go:107]:
```suggestion
					if id != nil && idConfig.RoleCertKeyFileOutput {
```

Not this??

Comment (WindzCUHK) [reply pkg/identity/certificated.go:107]: ONLY allow disable on single SIA pattern (internal key rotation).
`id == nil` =>  NO internal rotation => always `true`

### cases
1. non-copperargos, external rotation, always `true`
2. sidecar pattern (non-copper, issue role cert ONLY), another SIA in copperargos, always `true`
3. copperargos, single SIA, allow `true/false`

**Comment by mlajkim** [comment]
memo
- [ ] athenz-sia.env?
```sh
#
# Outputs private key specifically for role certificates (e.g. :true or false) (default: true)
#
ROLE_CERT_KEY_FILE_OUTPUT=true
```

File: pkg/identity/certificated.go (hunk 1 above)
Comment (mlajkim) [reply pkg/identity/certificated.go:107]: @WindzCUHK Ah, makes sense. Thank you.

Comment (mlajkim) [reply pkg/identity/certificated.go:107]: Kinda wonder how end user would feel when they set `RoleCertKeyFileOutput` as `false` and still outputs the role key.

**Comment by WindzCUHK** [comment]
>  athenz-sia.env?
multiple option missing... Can we handle this later in the following issue?
https://github.com/AthenZ/k8s-athenz-sia/issues/40

File: pkg/identity/certificated.go (hunk 1 above)
Comment (WindzCUHK) [reply pkg/identity/certificated.go:107]: `RoleCertKeyFileOutput` => `RoleCertKeyFileOutputOnInternalKey`?

Comment (mlajkim) [reply pkg/identity/certificated.go:107]: @WindzCUHK 
Mmm lets just stick with `RoleCertKeyFileOutput`.


> Kinda wonder how end user would feel when they set RoleCertKeyFileOutput as false and still outputs the role key.

nvm

**Review by mlajkim — APPROVED** [review]
LGTM

**Comment by mlajkim** [comment]
The changes here happen to be destructive and changes will be handled in the issue [here](https://github.com/AthenZ/k8s-athenz-sia/issues/42).

**Pull request merged**